klippy/extras/resonance_tester.py

```diff
@@ -0,0 +1,283 @@
+                        raise gcmd.error(
+                                "%s-axis accelerometer measured no data" % (
+                                    chip_axis,))
+                    new_data = helper.process_accelerometer_data(chip_values)
+                    if calibration_data[axis] is None:
+                        calibration_data[axis] = new_data
+                    else:
+                        calibration_data[axis].join(new_data)
+
+        configfile = self.printer.lookup_object('configfile')
+
+        for axis in calibrate_axes:
+            gcmd.respond_info(
+                    "Calculating the best input shaper parameters for %s axis"
+                    % (axis,))
+            calibration_data[axis].normalize_to_frequencies()
+            shaper_name, shaper_freq, shapers_vals = helper.find_best_shaper(
+                    calibration_data[axis], gcmd.respond_info)
+            gcmd.respond_info(
+                    "Recommended shaper_type_%s = %s, shaper_freq_%s = %.1f Hz"
+                    % (axis, shaper_name, axis, shaper_freq))
+            helper.save_params(configfile, axis, shaper_name, shaper_freq)
+            self.save_calibration_data(csv_name_tmpl, helper, axis,
+                                       calibration_data[axis], shapers_vals)
+
+        gcmd.respond_info(
+            "The SAVE_CONFIG command will update the printer config file\n"
+            "with these parameters and restart the printer.")
+        if input_shaper is not None:
+            input_shaper.enable_shaping()
+            gcmd.respond_info("Re-enabled [input_shaper] after calibration")
+
+    def cmd_MEASURE_AXES_NOISE(self, gcmd):
+        meas_time = gcmd.get_float("MEAS_TIME", 2.)
+        for _, chip in self.accel_chips:
+            chip.start_measurements()
+        self.printer.lookup_object('toolhead').dwell(meas_time)
+        raw_values = [(axis, chip.finish_measurements())
+                      for axis, chip in self.accel_chips]
+        helper = shaper_calibrate.ShaperCalibrate(self.printer)
+        for axis, raw_data in raw_values:
+            data = helper.process_accelerometer_data(raw_data)
+            vx = data.psd_x.mean()
+            vy = data.psd_y.mean()
+            vz = data.psd_z.mean()
+            gcmd.respond_info("Axes noise for %s-axis accelerometer: "
+                              "%.6f (x), %.6f (y), %.6f (z)" % (
+                                  axis, vx, vy, vz))
+
+    def get_filename(self, name_tmpl, axis=None, point=None):
+        # Cleanup name_tmpl and get only the name
+        name_tmpl = os.path.split(os.path.normpath(name_tmpl))[1]
+        base, ext = os.path.splitext(name_tmpl)
+        time_suffix = time.strftime("_%Y%m%d_%H%M%S")
+        if axis:
+            base += '_' + axis
+        if point:
+            base += "_%.3f_%.3f" % (point[0], point[1])
+        base += time_suffix
+        return os.path.join("/tmp", base + ext)
```

Comment:
> I'm leery of allowing users to specify filenames from the g-code command line - both because I'm leery of an accident overriding an important system file and because I'm leery of a malicious gcode file being used to corrupt a system.

Note that the code currently runs `os.path.split(os.path.normpath(name_tmpl))[1]` to at least attempt to prevent a user from doing something weird and writing files outside of `/tmp/` folder. And in terms of malicious gcode: I honestly think that the rouge gcode can do much more damage than just overwriting an important file and corrupting a system: it can cause a physical damage to the printer and even maybe set it on fire. So in general, it is a very bad idea to take untrusted gcode and run it on your printer. Though I agree that sometimes it is necessary to do that for debugging.   

> As such, I'd prefer to enforce a file name template determined by the source code - for example `/tmp/calibration_data_<name>.csv` where the user may only specify `<name>` and the code verifies that `<name>` does not contain any weird characters. For example, see how the adxl345.py code performs this check.

In general, forcing a fixed template with some short <name> override might be a good idea though, because it may also simplify the documentation how it is supposed to work.

Reply:
> Note that the code currently runs os.path.split(os.path.normpath(name_tmpl))[1]

Ah, I saw that, but wasn't sure what it did.  FWIW, I'd also prevent "weird" characters in the filename like '$', '`', and similar.

I agree that running malicious g-code would be very dangerous.  However, I would still feel better if we didn't add "system level privilege escalation" to the list of dangers.  I guess I don't see a real-world use case for "custom filenames" and I'd just feel better if we didn't "open that can of worms".

Reply:
> However, I would still feel better if we didn't add "system level privilege escalation" to the list of dangers. I guess I don't see a real-world use case for "custom filenames" and I'd just feel better if we didn't "open that can of worms".

OK, makes sense, I made changes to the output naming scheme similar to ACCELEROMETER_MEASURE so this should not longer be an issue.